File: src/main/java/org/hisp/dhis/lib/expression/spi/QueryModifiers.java
package org.hisp.dhis.lib.expression.spi;

import lombok.Builder;
import lombok.Value;
import org.hisp.dhis.lib.expression.ast.AggregationType;

import java.time.LocalDate;
import java.util.function.BiConsumer;

@Value
@Builder(toBuilder = true)
public class QueryModifiers {

    /**
     * Use aggregation over periods when loading data.
     * Value then must be a {@code double[]}.
     */
    boolean periodAggregation;
    AggregationType aggregationType;
    LocalDate maxDate;
    LocalDate minDate;
    Integer periodOffset;
    Integer stageOffset;
    boolean yearToDate;
    String subExpression; // SQL

    @Override
    public String toString() {
        StringBuilder str = new StringBuilder();
        BiConsumer<String,Object> toStr = (name, value) -> {
            if (value != null && value != Boolean.FALSE) {
                str.append(".").append(name).append("(").append(value == Boolean.TRUE ? "" : value.toString()).append(")");
            }
        };
        toStr.accept("periodAggregation", periodAggregation);
        toStr.accept("aggregationType", aggregationType);
        toStr.accept("maxDate", maxDate);
        toStr.accept("minDate", minDate);
        toStr.accept("periodOffset", periodOffset);
        toStr.accept("stageOffset", stageOffset);
        toStr.accept("yearToDate", yearToDate);
        return str.toString();
    }
}

File: src/jvmMain/java/org/hisp/dhis/lib/expression/spi/QueryModifiers.kt
package org.hisp.dhis.lib.expression.spi

import org.hisp.dhis.lib.expression.ast.AggregationType
import java.time.LocalDate

data class QueryModifiers (
    /**
     * Use aggregation over periods when loading data. Value then must be a `double[]`.
     */
    val periodAggregation: Boolean,
    val aggregationType: AggregationType?,
    val maxDate: LocalDate?,
    val minDate: LocalDate?,
    val periodOffset: Int?,
    val stageOffset: Int?,
    val yearToDate: Boolean,
    val subExpression: String? // SQL
){

    constructor() : this(false, null, null, null, null, null, false, null)

    override fun toString(): String {
        val str = StringBuilder()
        val toStr = { name: String?, value: Any? ->
            if (value != null) {
                if (value is Boolean) {
                    if (value == true) str.append(".").append(name).append("()");
                } else {
                    str.append(".").append(name).append("(").append(value.toString()).append(")")
                }
            }
        }
        toStr("periodAggregation", periodAggregation)
        toStr("aggregationType", aggregationType)
        toStr("maxDate", maxDate)
        toStr("minDate", minDate)
        toStr("periodOffset", periodOffset)
        toStr("stageOffset", stageOffset)
        toStr("yearToDate", yearToDate)
        return str.toString()
    }
}
